docs/src/tutorial/cython_tutorial.rst

```diff
@@ -127,91 +130,142 @@ Primes
-and will be copied into a Python list at the end (line 22).
+        .. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.pyx
+            :lines: 2,3
+            :dedent:
+            :lineno-start: 2
+
+        Lines 2 and 3 use the ``cdef`` statement to define some local C variables.
+        The result is stored in the C array ``p`` during processing,
+        and will be copied into a Python list at the end (line 26).
 
 .. NOTE:: You cannot create very large arrays in this manner, because
-          they are allocated on the C function call :term:`stack<Stack allocation>`, which is a
-          rather precious and scarce resource.
+          they are allocated on the C function call :term:`stack<Stack allocation>`,
+          which is a rather precious and scarce resource.
           To request larger arrays,
           or even arrays with a length only known at runtime,
           you can learn how to make efficient use of
           :ref:`C memory allocation <memory_allocation>`,
           :ref:`Python arrays <array-array>`
           or :ref:`NumPy arrays <memoryviews>` with Cython.
 
-::
-
-    if nb_primes > 1000:
-        nb_primes = 1000
+.. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.pyx
+    :lines: 5,6
+    :dedent:
+    :lineno-start: 5
 
 As in C, declaring a static array requires knowing the size at compile time.
 We make sure the user doesn't set a value above 1000 (or we would have a
-segmentation fault, just like in C).  ::
+segmentation fault, just like in C)
+
+.. tabs::
+    .. group-tab:: Pure Python
+
+        .. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.py
+            :lines: 8,9
+            :dedent:
+            :lineno-start: 8
 
-    len_p = 0  # The number of elements in p
-    n = 2
-    while len_p < nb_primes:
+        If we run the code from python, then we must initialize the array,
+        the easiest way is to fill it with zeros as on lines 8-9.
+        If we compile this, then the array will already be filled with
+        random data from the allocated memory, since we will change them anyway,
+        we do not need to overwrite this random data.
 
-Lines 7-9 set up for a loop which will test candidate numbers for primeness
-until the required number of primes has been found. ::
+        .. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.py
+            :lines: 10-13
+            :dedent:
+            :lineno-start: 10
 
-    # Is n prime?
-    for i in p[:len_p]:
-        if n % i == 0:
-            break
+    .. group-tab:: Cython
 
-Lines 11-12, which try dividing a candidate by all the primes found so far,
+        .. literalinclude:: ../../examples/tutorial/cython_tutorial/primes.pyx
+            :lines: 10-13
+            :dedent:
+            :lineno-start: 10
+
+Lines 11-13 set up for a loop which will test candidate numbers for primeness
+until the required number of primes has been found.
```

Comment:
It is rather confusing to call it a for loop.
```suggestion
Lines 11-13 sets up a while loop that starts looping through all numbers looking for primes,
stopping when the required number of primes have been found.
```

Reply:
"Loop-looping" doesn't sound very good.

Ok, [here](https://github.com/cython/cython/pull/4226/commits/5a44abb98f64e723283c6460d119a90979a8c899) is some mix, I think it worked out quite well.